build-data-capturing-gradle-samples/capture-thermal-throttling/gradle-thermal-throttling.gradle

```diff
@@ -0,0 +1,149 @@
+import org.gradle.api.services.BuildService
+import org.gradle.api.services.BuildServiceParameters
+import org.gradle.internal.os.OperatingSystem
+
+import java.nio.charset.Charset
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.function.Function
+
+import org.gradle.tooling.events.OperationCompletionListener
+import org.gradle.tooling.events.FinishEvent
+
+/**
+ * This Gradle script captures the thermal throttling level and adds it as a tag.
+ * Some parameters can be tweaked according to your build:
+ * - SAMPLING_INTERVAL_IN_SECONDS, frequency at which the capture command is run
+ * - THROTLLING_LEVEL, list of throttling levels and value ranges (to be compared with captured throttling average value)
+ *
+ * WARNINGS:
+ * - This is supported on MacOS only.
+ *
+ * This should be applied to the root project:
+ * <code> apply from: file('gradle-thermal-throttling.gradle') </code>
+ */
+
+if (OperatingSystem.current().macOsX) {
+  def buildScanApi = project.extensions.findByName('buildScan')
+  if (!buildScanApi) {
+    return
+  }
+
+  // Register Thermal throttling service
+  def throttlingServiceProvider = gradle.sharedServices.registerIfAbsent("thermalThrottling", ThermalThrottlingService.class) {}
+
+  // start service
+  gradle.services.get(BuildEventsListenerRegistry).onTaskCompletion(throttlingServiceProvider)
+
+  buildScan.buildFinished { def result ->
+    try (def service = throttlingServiceProvider.get()) {
+      // process results on build completion
+      service.processResults(buildScanApi)
+    }
+  }
+} else {
+  println "INFO - Not running on MacOS - no thermal throttling data will be captured"
+}
+
+// Thermal Throttling service implementation
+abstract class ThermalThrottlingService implements BuildService<BuildServiceParameters.None>, AutoCloseable, OperationCompletionListener {
```

Comment:
Very interesting that we can use the `OperationCompletionListener` to mark the `BuildService` lifecycle to work with CC. Is there a bug filed somewhere for this?

Reply:
This is the way it works afaik, I leave it to @mlopatkin to confirm

Reply:
Revival of registered `OperationCompletionListener`s is a supported feature, though the intended use case is for actual listeners. Anyway, currently we don't have anything better for service bootstrapping and this API isn't going away soon, so using it looks fine to me.